File: football-api/app/services/match_listing.py
"""Regras de classificação 'partida atual' compartilhadas entre routers.

Uma partida é considerada `is_current` quando ainda merece destaque no card
de "Atuais" do dashboard / aba do grupo. A regra combina três sinais:

1. Status `open` ou `in_progress` (jogo ainda por vir / acontecendo).
2. Janela de votação não fechou (`voting_status != 'closed'`).
3. É a partida mais recente do grupo E não há partida futura criada
   ainda — evita que a aba "Atuais" fique vazia entre rachões.

Documentado em docs/prd/044 §17 (paridade v2→prod oficial) — a v2 (Go) deve
expor os mesmos campos quando atender o frontend de produção.
"""

from collections.abc import Iterable

from app.models.match import MatchStatus
from app.services.voting import voting_status as compute_voting_status


def classify_matches(matches: Iterable) -> dict[str, tuple[bool, str]]:
    """Para cada match na coleção, retorna (is_current, voting_status).

    `matches` deve ser a lista completa de partidas DO MESMO grupo — o cálculo
    do "é a mais recente sem próxima criada" precisa do contexto do grupo. Se
    matches vier de múltiplos grupos, agrupe antes (vide players.get_my_matches).
    """
    matches_list = list(matches)
    if not matches_list:
        return {}

    has_future = any(
        m.status in (MatchStatus.OPEN, MatchStatus.IN_PROGRESS) for m in matches_list
    )
    closed_matches = [m for m in matches_list if m.status == MatchStatus.CLOSED]
    most_recent_closed_id = None
    if closed_matches:
        # Ordena por (match_date, start_time) descendente — pega a última do grupo.
        most_recent_closed_id = max(
            closed_matches, key=lambda m: (m.match_date, m.start_time)
        ).id

    out: dict[str, tuple[bool, str]] = {}
    for m in matches_list:
        vstatus = compute_voting_status(m)
        if m.status in (MatchStatus.OPEN, MatchStatus.IN_PROGRESS):
            is_current = True
        elif vstatus in ("not_open", "open"):
            is_current = True
        elif m.id == most_recent_closed_id and not has_future:
            is_current = True
        else:
            is_current = False
        out[str(m.id)] = (is_current, vstatus)
    return out

File: football-api-go/internal/services/match_listing.go
// Package services — match listing classifier (mirror of
// football-api/app/services/match_listing.py).
//
// Decide se cada partida pertence ao card "Atuais" do frontend, usando 3 sinais:
//  1. status `open` ou `in_progress` (jogo por vir / acontecendo).
//  2. janela de votação ainda não fechou.
//  3. é a partida mais recente do grupo E não há partida futura criada ainda
//     (evita que a aba "Atuais" fique vazia entre rachões).
//
// PRD 044 §17 — item de paridade entregue.
package services

import (
	"github.com/google/uuid"
)

// ListingMatch é o subset mínimo de campos necessários pra classificar.
// Tanto db.Match quanto db.PlayerMatch podem fornecer esses dados.
type ListingMatch struct {
	ID                   uuid.UUID
	Status               string // "open" | "in_progress" | "closed"
	MatchDate            string // "YYYY-MM-DD"
	StartTime            string // "HH:MM:SS"
	EndTime              *string
	VoteOpenDelayMinutes int
	VoteDurationHours    int
}

// ClassificationResult armazena os dois campos derivados por partida.
type ClassificationResult struct {
	IsCurrent     bool
	VotingStatus  VotingStatus
}

// ClassifyMatches recebe a lista completa de partidas DE UM MESMO GRUPO e
// devolve um mapa MatchID → (IsCurrent, VotingStatus). Se a lista cruzar
// múltiplos grupos, agrupe antes de chamar (ver handlers/players.go).
func ClassifyMatches(matches []ListingMatch) map[uuid.UUID]ClassificationResult {
	out := make(map[uuid.UUID]ClassificationResult, len(matches))
	if len(matches) == 0 {
		return out
	}

	hasFuture := false
	for _, m := range matches {
		if m.Status == "open" || m.Status == "in_progress" {
			hasFuture = true
			break
		}
	}

	// Determinar a partida fechada mais recente (por match_date, start_time desc).
	var mostRecentClosedID uuid.UUID
	var mostRecentClosedKey string
	for _, m := range matches {
		if m.Status != "closed" {
			continue
		}
		key := m.MatchDate + "T" + m.StartTime
		if key > mostRecentClosedKey {
			mostRecentClosedKey = key
			mostRecentClosedID = m.ID
		}
	}

	for _, m := range matches {
		vstatus := ComputeVotingStatus(VotingInput{
			MatchDate:            m.MatchDate,
			StartTime:            m.StartTime,
			EndTime:              m.EndTime,
			VoteOpenDelayMinutes: m.VoteOpenDelayMinutes,
			VoteDurationHours:    m.VoteDurationHours,
		})

		isCurrent := false
		switch {
		case m.Status == "open" || m.Status == "in_progress":
			isCurrent = true
		case vstatus == VotingStatusNotOpen || vstatus == VotingStatusOpen:
			isCurrent = true
		case m.ID == mostRecentClosedID && !hasFuture:
			isCurrent = true
		}

		out[m.ID] = ClassificationResult{
			IsCurrent:    isCurrent,
			VotingStatus: vstatus,
		}
	}
	return out
}
